miniapps/spde/material_metrics.hpp

```diff
@@ -0,0 +1,102 @@
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details
+
+#ifndef MATERIAL_METRICS_HPP
+#define MATERIAL_METRICS_HPP
+
+#include "mfem.hpp"
+#include <random>
+#include <vector>
+
+namespace mfem {
+
+/// Class that implements an edge defined by a start and end point.
+class Edge {
+public:
+  Edge(const Vector &start, const Vector &end) : start_(start), end_(end) {}
+
+  /// Compute the distance between a point and the edge.
+  double GetDistanceTo(const Vector &x) const;
+
+private:
+  const Vector &start_;
+  const Vector &end_;
+};
+
+/// Virtual class to define the interface for defining the material topology.
+class MaterialTopology {
+public:
+  virtual ~MaterialTopology() = default;
+
+  /// Compute the metric rho describing the material topology.
+  virtual double ComputeMetric(const Vector &x) = 0;
+};
+
+/// Class that implements the particle topology.
+class ParticleTopology : public MaterialTopology {
+public:
+  /// Constructor. The length of the vectors random_positions and
+  /// random_rotations must be 3 * number_of_particles and
+  /// 9 * number_of_particles, respectively. They implicitly define the the
+  /// number of the particles.
+  /// @param[in]  (length_x, length_y, length_z) - particle shape
+  /// @param[in]  random_positions - vector with random positions for partices
+  /// @param[in]  random_rotations - vector with random rotations for particles
+  ParticleTopology(double length_x, double length_y, double length_z,
+                   std::vector<double> &random_positions,
+                   std::vector<double> &random_rotations)
+      : number_of_particles_(random_positions.size() / 3),
+        particle_shape_({length_x, length_y, length_z}) {
+    Initialize(random_positions, random_rotations);
+  }
+
+  /// Compute the metric rho describing the particle topology. For a vector x,
+  /// this function returns the shortest distance to any of the particles. The
+  /// individual is computed as || A_k (x-x_k) ||_2. (A allows do distort the
+  /// particle shape.)
+  double ComputeMetric(const Vector &x) final;
+
+private:
+  /// Initialize the particle topology with positions x_k and matrices A_k.
+  void Initialize(std::vector<double> &random_positions,
+                  std::vector<double> &random_rotations);
+
+  std::vector<Vector> particle_positions_;         // A_k * x_k, scaled!
+  std::vector<DenseMatrix> particle_orientations_; // Random rotations of shape
+  Vector particle_shape_;                          // The shape of the particle.
+  int number_of_particles_;                        // The number of particles.
```

Comment:
The order of these definitions does not match the order of the initializers on lines 56 and 57. This is generating compiler warnings in `clang`.